File: hang.rb
require 'arrow'

100.times do |index|
  puts index
  plan = Arrow::ExecutePlan.new
  left_table = Arrow::Table.load('foo.arrow')
  node = plan.build_source_node(left_table)
  %w[bar baz qux].each do |right_name|
    right_table = Arrow::Table.load("#{right_name}.arrow")
    right_node = plan.build_source_node(right_table)
    join_options = Arrow::HashJoinNodeOptions.new(:left_outer, [right_name], [right_name])
    node = plan.build_hash_join_node(node, right_node, join_options)
  end
  sink_node_options = Arrow::SinkNodeOptions.new
  plan.build_sink_node(node, sink_node_options)
  plan.validate
  plan.start
  plan.wait
end

File: hang.py
import pyarrow as pa
import pyarrow.acero as acero

for index in range(100):
    print(index)
    left_table = pa.table(pa.ipc.open_file("foo.arrow").read_all())
    node = acero.Declaration("table_source", acero.TableSourceNodeOptions(left_table))
    for right_name in ["bar", "baz", "qux"]:
        right_table = pa.table(pa.ipc.open_file(f"{right_name}.arrow").read_all())
        right_node = acero.Declaration(
            "table_source", acero.TableSourceNodeOptions(right_table)
        )
        join_options = acero.HashJoinNodeOptions(
            join_type="left outer", left_keys=[right_name], right_keys=[right_name]
        )
        node = acero.Declaration("hashjoin", join_options, [node, right_node])
    node.to_table()
